front/pages/api/lookup.ts

```diff
@@ -0,0 +1,174 @@
+      api_error: {
+        type: "invalid_basic_authorization_error",
+        message: "Invalid token",
+      },
+    });
+  }
+  const { resource } = req.query;
+  const resourceValidation = ResourceType.decode(resource);
+  if (isLeft(resourceValidation)) {
+    return apiError(req, res, {
+      status_code: 400,
+      api_error: {
+        type: "invalid_request_error",
+        message: "Invalid resource type. Must be 'user' or 'workspace'",
+      },
+    });
+  }
+
+  let response: LookupResponseBody | null = null;
+  switch (resourceValidation.right) {
+    case "user":
+      {
+        const bodyValidation = UserLookupSchema.decode(req.body);
+        if (isLeft(bodyValidation)) {
+          const pathError = reporter.formatValidationErrors(
+            bodyValidation.left
+          );
+          return apiError(req, res, {
+            status_code: 400,
+            api_error: {
+              type: "invalid_request_error",
+              message: `Invalid request body for user lookup: ${pathError}`,
+            },
+          });
+        }
+        response = await handleLookupUser(bodyValidation.right);
+      }
+      break;
+    case "workspace": {
+      const bodyValidation = WorkspaceLookupSchema.decode(req.body);
+      if (isLeft(bodyValidation)) {
+        const pathError = reporter.formatValidationErrors(bodyValidation.left);
+        return apiError(req, res, {
+          status_code: 400,
+          api_error: {
+            type: "invalid_request_error",
+            message: `Invalid request body for user lookup ${pathError}`,
+          },
+        });
+      }
+      response = await handleLookupWorkspace(bodyValidation.right);
+    }
+  }
+  res.status(200).json(response);
+  return;
+}
+
+async function handleLookupUser(
+  userLookup: t.TypeOf<typeof UserLookupSchema>
+): Promise<UserLookupResponse> {
+  // Check if user exists
+  const user = await fetchUserWithAuth0Sub(userLookup.user.sub);
+
+  // Check for pending invitations
+  const pendingInvite =
+    await getPendingMembershipInvitationWithWorkspaceForEmail(
+      userLookup.user.email
+    );
+
+  // Check for workspace with verified domain
+  const workspaceWithVerifiedDomain = await findWorkspaceWithVerifiedDomain({
+    email: userLookup.user.email,
+    email_verified: userLookup.user.email_verified,
+  });
+
+  // Check auto-join
+  const canAutoJoin =
+    workspaceWithVerifiedDomain?.domainAutoJoinEnabled ?? false;
+
+  return {
+    status: user ? "member" : pendingInvite ? "invited" : "new",
+    workspace: {
+      autoJoin: canAutoJoin,
+      id:
+        pendingInvite?.workspace.sId ||
+        (canAutoJoin ? workspaceWithVerifiedDomain?.workspace?.sId : undefined),
+    },
+  };
+}
+
+async function handleLookupWorkspace(
+  body: t.TypeOf<typeof WorkspaceLookupSchema>
+): Promise<WorkspaceLookupResponse> {
+  const workspace = await Workspace.findOne({
+    where: { sId: body.workspace },
+  });
+  return {
+    isNew: !!workspace,
```

Comment:
Let's return:

```
{
  workspace: null | { sId: string }
}
```

There is value in being minimalistic here to not leak any data but we still need to be principled in the value we return